File: solutions/haskell/space-age/1/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune
  deriving Eq

earthlyPeriodInSeconds :: Float
earthlyPeriodInSeconds = 31557600

ageOn :: Planet -> Float -> Float
ageOn planet seconds = seconds / (earthlyPeriodInSeconds * periodInEarthYears) where
  periodInEarthYears = case planet of
                           Mercury  ->  0.2408467
                           Venus    ->  0.61519726
                           Earth    ->  1.0
                           Mars     ->  1.8808158
                           Jupiter  ->  11.862615
                           Saturn   ->  29.447498
                           Uranus   ->  84.016846
                           Neptune  ->  164.79132

File: solutions/javascript/space-age/1/space-age.js
class OrbitalPeriod {
  static get Mercury() {
    return 0.2408467;
  }

  static get Venus() {
    return 0.61519726;
  }

  static get Earth() {
    return 1.0;
  }

  static get Mars() {
    return 1.8808158;
  }

  static get Jupiter() {
    return 11.862615;
  }

  static get Saturn() {
    return 29.447498;
  }

  static get Uranus() {
    return 84.016846;
  }

  static get Neptune() {
    return 164.79132;
  }
}

function secondsToYears(seconds) {
  const secondsInYear = 60 * 60 * 24 * 365.25;
  return seconds / secondsInYear;
}

/**
  * params {string} planet
  * params {number} seconds
  * @returns {number}
  */
export const age = (planet, seconds) => {
  const normalizedPlanet = planet.trim().replace(/^\w/, (letter) => letter.toUpperCase());
  const orbitalPeriod = OrbitalPeriod[normalizedPlanet];
  if (orbitalPeriod === undefined) throw new Error("not a planet");
  
  const earthYears = secondsToYears(seconds);
  const yearsInPlanet = earthYears / orbitalPeriod;
  
  return Math.round(100 * yearsInPlanet) / 100; 
};
